File: python/infix_to_postfix_converter.py
# https://www.codewars.com/kata/52e864d1ffb6ac25db00017f/train/python

import re

def to_postfix(infix):
    ops, stack, out = {'+': 1, '-': 1, '*': 2, '/': 2, '^': 3}, [], []
    for token in re.findall(r'\d+|[+\-*/^()]', infix):
        if token.isdigit(): out.append(token)
        elif token == '(': stack.append(token)
        elif token == ')':
            while stack and stack[-1] != '(': out.append(stack.pop())
            stack.pop()
        else:
            while stack and ops[token] <= ops.get(stack[-1], 0) and token != '^': out.append(stack.pop())
            stack.append(token)
    return ''.join(out + stack[::-1])

File: js/infix_to_postfix_converter.js
// https://www.codewars.com/kata/52e864d1ffb6ac25db00017f/train/javascript

function toPostfix(infix) {
    const ops = {'+': 1, '-': 1, '*': 2, '/': 2, '^': 3}, stack = [], out = [];
    infix.replace(/\d+|[+\-*/^()]/g, token => {
      if (!isNaN(token)) out.push(token); 
      else if (token === '(') stack.push(token); 
      else if (token === ')') { 
        while (stack.length && stack.at(-1) !== '(') out.push(stack.pop());
        stack.pop(); 
      } else { 
        while (stack.length && ops[token] <= ops[stack.at(-1)] && token !== '^') out.push(stack.pop());
        stack.push(token);
      }
    });
    while (stack.length) out.push(stack.pop());
    return out.join('');
  }